zcash_client_sqlite/src/wallet/scanning.rs

```diff
@@ -0,0 +1,1089 @@
+    }
+}
+
+#[derive(Debug, Clone)]
+enum SpanningTree {
+    Leaf(ScanRange),
+    Parent {
+        span: Range<BlockHeight>,
+        left: Box<SpanningTree>,
+        right: Box<SpanningTree>,
+    },
+}
+
+impl SpanningTree {
+    fn span(&self) -> Range<BlockHeight> {
+        match self {
+            SpanningTree::Leaf(entry) => entry.block_range().clone(),
+            SpanningTree::Parent { span, .. } => span.clone(),
+        }
+    }
+
+    fn from_joined(joined: Joined) -> Self {
+        match joined {
+            Joined::One(entry) => SpanningTree::Leaf(entry),
+            Joined::Two(left, right) => SpanningTree::Parent {
+                span: left.block_range().start..right.block_range().end,
+                left: Box::new(SpanningTree::Leaf(left)),
+                right: Box::new(SpanningTree::Leaf(right)),
+            },
+            Joined::Three(left, mid, right) => SpanningTree::Parent {
+                span: left.block_range().start..right.block_range().end,
+                left: Box::new(SpanningTree::Leaf(left)),
+                right: Box::new(SpanningTree::Parent {
+                    span: mid.block_range().start..right.block_range().end,
+                    left: Box::new(SpanningTree::Leaf(mid)),
+                    right: Box::new(SpanningTree::Leaf(right)),
+                }),
+            },
+        }
+    }
+
+    fn from_insert(
+        left: Box<Self>,
+        right: Box<Self>,
+        to_insert: ScanRange,
+        insert: Insert,
+    ) -> Self {
+        let (left, right) = match insert {
+            Insert::Left => (Box::new(left.insert(to_insert)), right),
+            Insert::Right => (left, Box::new(right.insert(to_insert))),
+        };
+        SpanningTree::Parent {
+            span: left.span().start..right.span().end,
+            left,
+            right,
+        }
+    }
+
+    fn from_split(left: Self, right: Self, to_insert: ScanRange, split_point: BlockHeight) -> Self {
+        let (l_insert, r_insert) = to_insert
+            .split_at(split_point)
+            .expect("Split point is within the range of to_insert");
+        let left = Box::new(left.insert(l_insert));
+        let right = Box::new(right.insert(r_insert));
+        SpanningTree::Parent {
+            span: left.span().start..right.span().end,
+            left,
+            right,
+        }
+    }
+
+    fn insert(self, to_insert: ScanRange) -> Self {
+        match self {
+            SpanningTree::Leaf(cur) => Self::from_joined(insert(cur, to_insert)),
+            SpanningTree::Parent { span, left, right } => {
+                // TODO: this algorithm always preserves the existing partition point, and does not
+                // do any rebalancing or unification of ranges within the tree; `into_vec`
+                // performs such unification, and the tree being unbalanced should be fine given
+                // the relatively small number of ranges we should ordinarily be concerned with.
```

Comment:
TODO comments should say what is to be done. In this case I think it's not necessary to do anything, so:
```suggestion
                // This algorithm always preserves the existing partition point, and does not do
                // any rebalancing or unification of ranges within the tree. This should be okay
                // because `into_vec` performs such unification, and the tree being unbalanced
                // should be fine given the relatively small number of ranges we should ordinarily
                // be concerned with.
```